lib/rules/media-query-no-invalid/__tests__/index.mjs

```diff
@@ -282,3 +282,67 @@ testRule({
 		},
 	],
 });
+
+testRule({
+	ruleName,
+	config: [true, { ignoreFunctions: ['theme', 'fooFunction', '/^get.*$/'] }],
+
+	accept: [
+		{
+			code: '@media (min-width: theme(screens.md)) {}',
+		},
+		{
+			code: '@media (min-width: theme(screens.md) and fooFunction(300px)) {}',
+		},
+		{
+			code: '@media (min-width: getDefaultWidth()) {}',
+		},
+		{
+			code: '@media (min-width: theme(screens.md) and var(300px)) {}',
+		},
+		{
+			code: '@media (min-width: var(300px) or theme(screens.md)) {}',
```

Comment:
question

@ybiquitous asked me to take a last look.
Based on:

> […such as in `@media` rules, where the `var()` function would not be valid.](https://drafts.csswg.org/css-env-1/#intro)

@romainmenke should we add a `customSyntax` or use `env()` instead?
e.g. `@custom-media --foo-bar (min-width: 300px);`

Reply:
I am not sure what you mean :)

Given the scope of this change, can you give some example CSS input, the ignore config and what you would expect to happen?

Reply:
Put simply, is `var(300px)` valid in this scope?

Reply:
From the perspective of "what is valid CSS":

No `var()` can't be used and there aren't any plans to change that.

From the perspective of this rule and this change:

You can now add config to ignore `var()`, which should silence any warnings.

[Current behavior demo]([Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIACBbGUCWBDABAChQwDsBaAdwygBcALeHANywCc8SSAzCCASh52AAdIoKoA6LhAHDRoyABsIzes3QBuGVQC+wrSAA0IDhnkwAcljQIQMAB6WADqbFgAzq4PgIRYwHNrQkQ4OII2tlQwRFCuofQA2sLBwaGuVACepvLEVCSQPhi+JKlYUSxQoYk4ALq6nnl+AGJKKFhU1gBWrt6esA4eiIHJIKkZMFlEVLEhIPKtMKmh+pUp6ZnZud5+RVQlUGVTobMRCyC1WkA))

<img width="302" alt="Screenshot 2024-11-09 at 12 53 47" src="https://github.com/user-attachments/assets/df68a73f-ccad-421e-8e28-78a7a71a36d4">

----

Still unsure what you mean or what your concern is :)

Reply:
> Still unsure what you mean or what your concern is :)

My concern is that at line 288 we have

> `config: [true, { ignoreFunctions: ['theme', 'fooFunction', '/^get.*$/'] }],`

which doesn't ignore `var` AFAIK?
Hence shouldn't l303–305 be part of `reject: []` instead?
Maybe I am missing something obvious.

Reply:
I've already tried explaining this multiple times :)

- https://github.com/stylelint/stylelint/pull/8060#discussion_r1807781359
- https://github.com/stylelint/stylelint/pull/8060#issuecomment-2466184844
- https://github.com/stylelint/stylelint/pull/8060#discussion_r1815551527

Anything that is unrecognized becomes `GeneralEnclosed`.
`GeneralEnclosed` literally means:
- try and fail to parse a [media feature](https://drafts.csswg.org/mediaqueries-5/#typedef-media-feature)
- find the nearest balanced closing parenthesis
- capture everything in a `GeneralEnclosed` node
- continue parsing

There is no way to recover from this.

We can't ignore the function itself because we were never validating functions to begin with.

We must ignore the nearest `GeneralEnclosed`.
Ignoring the nearest `GeneralEnclosed` literally means to ignore anything in that block.

For this input: `@media (min-width: var(300px) or theme(screens.md)) {}` the nearest `GeneralEnclosed` is `(min-width: var(300px) or theme(screens.md))`.

Because the `var(300px)` is also contained within that ignored `GeneralEnclosed` it is also ignored, even though it is not covered by the config.

This is not am implementation detail as you implied in https://github.com/stylelint/stylelint/pull/8060#issuecomment-2466187361

This is how the CSS specification defines media queries must be parsed and how their error recovery works.

This Stylelint rule intends to parse as a browser would and give feedback on parse errors.

Reply:
And to clarify why this test is correct and intended.
We want to have coverage for the behavior I just described.

This test adds that coverage.